File: src/swift/3_StructuralPatterns/5_Facade/Facade.swift
import Foundation
import XCTest

class Buffer
{
  var width, height: Int
  var buffer: [Character]
  init(_ width: Int, _ height: Int)
  {
    self.width = width
    self.height = height
    buffer = [Character](repeating: " ", count: 64)
  }

  subscript(_ index: Int) -> Character
  {
    return buffer[index]
  }
}

class Viewport
{
  var buffer: Buffer
  var offset = 0
  init(_ buffer: Buffer)
  {
    self.buffer = buffer
  }

  func getCharacterAt(_ index: Int) -> Character
  {
    return buffer[offset + index]
  }
}

class Console
{
  var buffers = [Buffer]()
  var viewports = [Viewport]()
  var offset = 0

  init()
  {
    let b = Buffer(10,10)
    let v = Viewport(b)
    buffers.append(b)
    viewports.append(v)
  }

  func getCharacterAt(_ index: Int) -> Character
  {
    // use default buffer
    return viewports[0].getCharacterAt(index)
  }
}

func main()
{
  let c = Console()
  let u = c.getCharacterAt(1)
}

main() // nothing to run here :|

File: src/go/3_StructuralPatterns/5_Facade/facade.go
package facade

type Buffer struct {
  width, height int
  buffer []rune
}

func NewBuffer(width, height int) *Buffer {
  return &Buffer { width, height,
    make([]rune, width*height)}
}

func (b *Buffer) At(index int) rune {
  return b.buffer[index]
}

type Viewport struct {
  buffer *Buffer
  offset int
}

func NewViewport(buffer *Buffer) *Viewport {
  return &Viewport{buffer: buffer}
}

func (v *Viewport) GetCharacterAt(index int) rune {
  return v.buffer.At(v.offset + index)
}

// a facade over buffers and viewports
type Console struct {
  buffers []*Buffer
  viewports []*Viewport
  offset int
}

func NewConsole() *Console {
  b := NewBuffer(10, 10)
  v := NewViewport(b)
  return &Console{[]*Buffer{b}, []*Viewport{v}, 0}
}

func (c *Console) GetCharacterAt(index int) rune {
  return c.viewports[0].GetCharacterAt(index)
}

func main() {
  c := NewConsole()
  u := c.GetCharacterAt(1)
}